pkg/quarterdeck/orgs.go

```diff
@@ -113,3 +119,80 @@ func (s *Server) ProjectCreate(c *gin.Context) {
 	project.Modified, _ = model.GetModified()
 	c.JSON(http.StatusOK, project)
 }
+
+func (s *Server) ProjectAccess(c *gin.Context) {
+	var (
+		err     error
+		exists  bool
+		project *api.Project
+		claims  *tokens.Claims
+		creds   *api.LoginReply
+	)
+
+	// Bind the Project request to the project data structure
+	if err = c.BindJSON(&project); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse request"))
+		return
+	}
+
+	// Validate the request from the API side.
+	if err = project.Validate(); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	// Fetch the user claims from the request
+	if claims, err = middleware.GetClaims(c); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("user claims unavailable"))
+		return
+	}
+
+	// Check that the orgID claim is allowed access to the project
+	model := &models.OrganizationProject{OrgID: claims.ParseOrgID(), ProjectID: project.ProjectID}
+	if exists, err = model.Exists(c.Request.Context()); !exists || err != nil {
+		if !exists {
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("unknown project id"))
+			return
+		}
+
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not complete project access request"))
+		return
+	}
+
+	// Prepare the credentials for the access token to make the one time access.
+	now := time.Now()
+	ota := &tokens.Claims{
+		RegisteredClaims: jwt.RegisteredClaims{
+			ID:        ulids.New().String(),
+			Subject:   claims.Subject,
+			IssuedAt:  jwt.NewNumericDate(now),
+			NotBefore: jwt.NewNumericDate(now),
+			ExpiresAt: jwt.NewNumericDate(now.Add(OneTimeAccessDuration)),
+		},
+		OrgID:       claims.OrgID,
+		ProjectID:   project.ProjectID.String(),
+		Permissions: make([]string, 0, 4),
+	}
+
+	// Add only the user permissions related to topics to these claims -- whatever
+	// access to topics the user has, so to will the one time access claims.
+	for _, permission := range claims.Permissions {
+		if strings.HasPrefix(permission, "topics") {
```

Comment:
Should we define the permission prefixes in code to protect ourselves from future permissions changes?

Reply:
Good idea